File: client/UiApp/Services/TelemetryService.cs
using System.Globalization;
using System.Net.Http;
using System.Net.Http.Json;
using System.Runtime.InteropServices;

namespace UiApp.Services;

/// <summary>
/// Sends periodic heartbeat to the signaling server for telemetry.
/// Collects: machine_id, app version, OS version, UI language.
/// All errors are swallowed — telemetry must never break the app.
/// </summary>
public sealed class TelemetryService : IDisposable
{
    private readonly HttpClient _http;
    private readonly string _baseUrl;
    private readonly string _machineId;
    private readonly string _appVersion;
    private readonly string _osVersion;
    private readonly string _osLanguage;
    private readonly Action<string>? _onLog;
    private readonly CancellationTokenSource _cts = new();
    private Task? _loop;

    public TelemetryService(HttpClient http, string baseUrl, string machineId, Action<string>? onLog = null)
    {
        _http = http;
        _baseUrl = baseUrl.TrimEnd('/');
        _machineId = machineId;
        _onLog = onLog;

        _appVersion = typeof(TelemetryService).Assembly.GetName().Version?.ToString(3) ?? "0.0.0";
        _osVersion = RuntimeInformation.OSDescription;
        _osLanguage = CultureInfo.CurrentUICulture.Name;
    }

    public void Start()
    {
        if (_loop is not null) return;
        _loop = Task.Run(() => HeartbeatLoopAsync(_cts.Token));
    }

    private async Task HeartbeatLoopAsync(CancellationToken ct)
    {
        // First heartbeat immediately on startup
        await SendHeartbeatAsync(ct);

        while (!ct.IsCancellationRequested)
        {
            try { await Task.Delay(TimeSpan.FromMinutes(5), ct); }
            catch (OperationCanceledException) { break; }

            await SendHeartbeatAsync(ct);
        }
    }

    private async Task SendHeartbeatAsync(CancellationToken ct)
    {
        try
        {
            var payload = new
            {
                machine_id = _machineId,
                app_version = _appVersion,
                os_version = _osVersion,
                os_language = _osLanguage
            };
            await _http.PostAsJsonAsync($"{_baseUrl}/api/v1/telemetry/heartbeat", payload, ct);
        }
        catch (OperationCanceledException) { /* shutdown */ }
        catch (Exception ex)
        {
            _onLog?.Invoke($"telemetry_heartbeat_error_{ex.GetType().Name}");
        }
    }

    public void Dispose()
    {
        _cts.Cancel();
        try { _loop?.Wait(TimeSpan.FromSeconds(2)); } catch { }
        _cts.Dispose();
    }
}

File: android/app/src/main/java/com/zconect/viewer/data/TelemetryService.kt
package com.zconect.viewer.data

import android.content.Context
import android.os.Build
import android.util.Log
import kotlinx.coroutines.*
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.Locale
import java.util.UUID
import java.util.concurrent.TimeUnit

/**
 * Sends periodic anonymous heartbeat to the signaling server.
 * Collects: machine_id (random UUID stored locally), platform, app version,
 * OS version, UI language, device model. GeoIP country resolved server-side.
 *
 * Privacy: NO personal data, NO IMEI/MAC/phone number, NO session credentials.
 * Errors are silently swallowed — telemetry MUST NOT break the app.
 */
class TelemetryService(private val context: Context) {

    companion object {
        private const val TAG = "Telemetry"
        private const val PREFS = "zconect_telemetry"
        private const val KEY_MACHINE_ID = "machine_id"
        private const val HEARTBEAT_INTERVAL_MIN = 5L // every 5 minutes
        private const val APP_VERSION = com.zconect.viewer.BuildConfig.VERSION_NAME
    }

    @Serializable
    private data class HeartbeatPayload(
        @SerialName("machine_id") val machineId: String,
        @SerialName("app_version") val appVersion: String,
        @SerialName("os_version") val osVersion: String,
        @SerialName("os_language") val osLanguage: String,
        val platform: String = "android",
        @SerialName("device_model") val deviceModel: String
    )

    private val json = Json { encodeDefaults = true; ignoreUnknownKeys = true }
    // Shared client (audit fix 2026-04-27).
    private val httpClient get() = HttpClientProvider.telemetry

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private var job: Job? = null

    /** Get or generate a persistent random machine_id. */
    private val machineId: String by lazy {
        val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        prefs.getString(KEY_MACHINE_ID, null) ?: run {
            val fresh = UUID.randomUUID().toString()
            prefs.edit().putString(KEY_MACHINE_ID, fresh).apply()
            fresh
        }
    }

    private val deviceModel: String by lazy {
        val m = Build.MANUFACTURER ?: ""
        val model = Build.MODEL ?: ""
        when {
            model.isBlank() -> m
            m.isBlank() -> model
            model.startsWith(m, ignoreCase = true) -> model
            else -> "$m $model"
        }.trim()
    }

    private val osVersion: String by lazy { "Android ${Build.VERSION.RELEASE}" }

    private val osLanguage: String by lazy {
        val locale = Locale.getDefault()
        val tag = locale.toLanguageTag()
        if (tag.isNullOrBlank() || tag == "und") "${locale.language}-${locale.country}" else tag
    }

    /** Start periodic heartbeat loop. Safe to call multiple times — only one loop runs. */
    fun start() {
        if (job?.isActive == true) return
        job = scope.launch {
            // First heartbeat right away
            sendHeartbeat()
            while (isActive) {
                delay(TimeUnit.MINUTES.toMillis(HEARTBEAT_INTERVAL_MIN))
                sendHeartbeat()
            }
        }
    }

    fun stop() {
        job?.cancel()
        job = null
    }

    private suspend fun sendHeartbeat() {
        try {
            val settings = AppSettings.getInstance(context)
            val baseUrl = settings.httpBaseUrl // "http://<server>"
            val payload = HeartbeatPayload(
                machineId = machineId,
                appVersion = APP_VERSION,
                osVersion = osVersion,
                osLanguage = osLanguage,
                deviceModel = deviceModel
            )
            val body = json.encodeToString(HeartbeatPayload.serializer(), payload)
            val request = Request.Builder()
                .url("$baseUrl/api/v1/telemetry/heartbeat")
                .post(body.toRequestBody("application/json".toMediaType()))
                .build()
            httpClient.newCall(request).execute().use { resp ->
                Log.d(TAG, "heartbeat response=${resp.code}")
            }
        } catch (e: Exception) {
            Log.w(TAG, "heartbeat failed: ${e.message}")
        }
    }
}
